### PR TITLE
[save as image] attach scope variables

### PR DESCRIPTION
**Pull request by nirvn**
## Description
@nyalldawson , this is needed to properly render symbols with expression-based styling rely on those.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [ ] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [ ] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [ ] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

### CODE REVIEW

File: src/app/qgisapp.cpp [hunk 1]
```diff
@@ -5854,6 +5854,15 @@ void QgisApp::saveMapAsImage()
     QgsMapSettings ms = QgsMapSettings();
     dlg.applyMapSettings( ms );
 
+    //build the expression context
+    QgsExpressionContext expressionContext;
+    expressionContext << QgsExpressionContextUtils::globalScope()
+                      << QgsExpressionContextUtils::projectScope( QgsProject::instance() )
+                      << QgsExpressionContextUtils::mapSettingsScope( ms )
+                      << new QgsExpressionContextScope();
```
Comment (nyalldawson) [inline src/app/qgisapp.cpp]: what's this empty scope for?

Comment (nirvn) [inline src/app/qgisapp.cpp]: Oh sorry, that's just me stupidly not cleaning up my copy/pasted code from QgsMapCanvas...

That said, should we attach the expression context in the QgsSaveAsDialog::applyMapSettings( QgsMapSettings & ) function? That'd de-duplicate the code.

Comment (nirvn) [inline src/app/qgisapp.cpp]: Err, QgsMapSaveDialog.

Comment (nyalldawson) [inline src/app/qgisapp.cpp]: Yes - sounds like a good idea to me

**Pull request merged**